File: fixtures/typescript-sample/src/user.service.ts
/**
 * User service for testing TypeScript extraction.
 */

export interface User {
  id: string;
  name: string;
  email: string;
  createdAt: Date;
  roles: string[];
}

export interface CreateUserDto {
  name: string;
  email: string;
  roles?: string[];
}

export class UserService {
  private users: Map<string, User> = new Map();

  /**
   * Create a new user.
   */
  createUser(dto: CreateUserDto): User {
    const id = crypto.randomUUID();
    const user: User = {
      id,
      name: dto.name,
      email: dto.email,
      createdAt: new Date(),
      roles: dto.roles ?? [],
    };
    this.users.set(id, user);
    return user;
  }

  /**
   * Get a user by ID.
   */
  getUser(id: string): User | undefined {
    return this.users.get(id);
  }

  /**
   * List all users.
   */
  listUsers(): User[] {
    return Array.from(this.users.values());
  }

  /**
   * Delete a user by ID.
   */
  deleteUser(id: string): boolean {
    return this.users.delete(id);
  }

  /**
   * Update a user.
   */
  updateUser(id: string, updates: Partial<CreateUserDto>): User | undefined {
    const user = this.users.get(id);
    if (!user) return undefined;

    const updated: User = {
      ...user,
      ...updates,
    };
    this.users.set(id, updated);
    return updated;
  }
}

export function createUserService(): UserService {
  return new UserService();
}

File: fixtures/python-sample/app/services/user_service.py
"""User service for testing RECON extraction."""

from dataclasses import dataclass, field
from typing import List, Optional
from datetime import datetime


@dataclass
class User:
    """User entity model."""
    id: str
    name: str
    email: str
    created_at: datetime = field(default_factory=datetime.now)
    roles: List[str] = field(default_factory=list)


class UserService:
    """Service for managing users."""

    def __init__(self):
        self._users: dict[str, User] = {}

    def create_user(self, id: str, name: str, email: str) -> User:
        """Create a new user."""
        user = User(id=id, name=name, email=email)
        self._users[id] = user
        return user

    def get_user(self, id: str) -> Optional[User]:
        """Get a user by ID."""
        return self._users.get(id)

    def list_users(self) -> List[User]:
        """List all users."""
        return list(self._users.values())

    def delete_user(self, id: str) -> bool:
        """Delete a user by ID."""
        if id in self._users:
            del self._users[id]
            return True
        return False


